chia/full_node/subscriptions.py

```diff
@@ -0,0 +1,137 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Dict, List, Optional, Set
+
+from chia.types.blockchain_format.sized_bytes import bytes32
+
+
+def _get_or_add_set(c: Dict[bytes32, Set[bytes32]], key: bytes32) -> Set[bytes32]:
+    ret: Optional[Set[bytes32]] = c.get(key)
+    if ret is not None:
+        return ret
+
+    ret2: Set[bytes32] = set()
+    c[key] = ret2
+    return ret2
+
+
+# The PeerSubscriptions class is essentially a multi-index container. It can be
+# indexed by peer_id, coin_id and puzzle_hash.
+@dataclass
```

Comment:
```suggestion
@dataclass(frozen=True)
```

Reply:
But this isn't frozen, it's entirely about mutation.  I guess it still blocks assignment of new dicts to the attributes...  I dunno.

Reply:
Its only about mutation of the members though, what is still possible when its frozen?

Reply:
Frozen only blocks reassignment of attributes, not access to them or modification of them.  When I think about frozen, I think of classes that have immutable attributes all the way down the tree.  A whole immutable tree of data once created.  Maybe that's too closed minded.

I guess frozen also blocks addition of new attributes, but slots can trigger that as well and mypy should protect against it generally.  Frozen also adjusts expectations around hashing where you'll get an exception about the dicts being unhashable instead of the instance of this class being unhashable.

Reply:
I must have misunderstood what frozen meant. I thought it meant that the object's value was immutable, but it doesn't seem to mean that. I'll add `frozen=True`